ffi/src/scan.rs

```diff
@@ -0,0 +1,367 @@
+    engine_visitor: extern "C" fn(
+        engine_context: *mut c_void,
+        engine_data: *mut EngineDataHandle,
+        selection_vector: KernelBoolSlice,
+    ),
+) -> ExternResult<bool> {
+    kernel_scan_data_next_impl(data, engine_context, engine_visitor)
+        .into_extern_result(data.engine_interface.error_allocator())
+}
+fn kernel_scan_data_next_impl(
+    data: &mut KernelScanDataIterator,
+    engine_context: *mut c_void,
+    engine_visitor: extern "C" fn(
+        engine_context: *mut c_void,
+        engine_data: *mut EngineDataHandle,
+        selection_vector: KernelBoolSlice,
+    ),
+) -> DeltaResult<bool> {
+    if let Some((data, sel_vec)) = data.data.next().transpose()? {
+        let bool_slice = KernelBoolSlice::from(sel_vec);
+        let data_handle = BoxHandle::into_handle(EngineDataHandle { data });
+        (engine_visitor)(engine_context, data_handle, bool_slice);
+        // ensure we free the data
+        unsafe { BoxHandle::drop_handle(data_handle) };
+        Ok(true)
+    } else {
+        Ok(false)
+    }
+}
+
+/// # Safety
+///
+/// Caller is responsible for (at most once) passing a valid pointer returned by a call to
+/// [kernel_scan_files_init].
+// we should probably be consistent with drop vs. free on engine side (probably the latter is more
+// intuitive to non-rust code)
+#[no_mangle]
+pub unsafe extern "C" fn kernel_scan_data_free(data: *mut KernelScanDataIterator) {
+    BoxHandle::drop_handle(data);
+}
+
+type CScanCallback = extern "C" fn(
+    engine_context: *mut c_void,
+    path: KernelStringSlice,
+    size: i64,
+    dv_info: *mut CDvInfo,
+    partition_map: *mut CStringMap,
+);
+
+pub struct CDvInfo {
+    dv_info: DvInfo,
+}
+impl BoxHandle for CDvInfo {}
+
+pub struct CStringMap {
+    values: HashMap<String, String>,
+}
+impl BoxHandle for CStringMap {}
+
+#[no_mangle]
+/// allow probing into a CStringMap. If the specified key is in the map, kernel will call
+/// allocate_fn with the value associated with the key and return the value returned from that
+/// function. If the key is not in the map, this will return NULL
+///
+/// # Safety
+///
+/// The engine is responsible for providing a valid [`CStringMap`] pointer and [`KernelStringSlice`]
+pub unsafe extern "C" fn get_from_map(
+    map: &mut CStringMap,
+    key: KernelStringSlice,
+    allocate_fn: AllocateStringFn,
+) -> *mut c_void {
+    let string_key = String::try_from_slice(key);
+    match map.values.get(&string_key) {
+        Some(v) => allocate_fn(v.as_str().into()),
+        None => std::ptr::null_mut(),
+    }
+}
+
+/// Get a selection vector out of a [`CDvInfo`] struct
+///
+/// # Safety
+/// Engine is responsible for providing valid pointers for each argument
+#[no_mangle]
+pub unsafe extern "C" fn selection_vector_from_dv(
+    info: &mut CDvInfo,
+    extern_engine_interface: *const ExternEngineInterfaceHandle,
+    state: &mut GlobalScanState,
+) -> ExternResult<*mut KernelBoolSlice> {
```

Comment:
Gah, github ate a longer comment I wrote to answer this.

In short, yes this could allocate a lot. There's a bigger question of how we want to present selection/deletion vectors to engines. We _could_ have a callback type thing here where we give them a `KernelBoolSlice*` and only promise it lives as long as the call, but it does seem wasteful to have allocated the whole slice which is valid in c land as well, and then just make them reallocate it all.

Ideally there'd be a quick way to turn a `RoaringTreemap` into whatever format the engine wanted. I'll ruminate a bit on an API that could enable that.